### ISSUE TITLE
How to save the predicted image?

### ISSUE THREAD

**Issue by LITDataScience** [lucidrains/DALLE2-pytorch#145]
I wanted to save the predicted image, tried various ways. Can you suggest some easy way?
```
import torch
from dalle2_pytorch import DALLE2, DiffusionPriorNetwork, DiffusionPrior, Unet, Decoder, OpenAIClipAdapter
import matplotlib.pyplot as plt
import numpy as np
import uuid

# openai pretrained clip - defaults to ViT-B/32

clip = OpenAIClipAdapter()

# mock data

text = torch.randint(0, 49408, (4, 256)).cuda()
images = torch.randn(4, 3, 256, 256).cuda()

# prior networks (with transformer)

prior_network = DiffusionPriorNetwork(dim=512, depth=6, dim_head=64, heads=8).cuda()

diffusion_prior = DiffusionPrior(net=prior_network, clip=clip, timesteps=100, cond_drop_prob=0.2).cuda()

loss = diffusion_prior(text, images)
loss.backward()

# do above for many steps ...

# decoder (with unet)

unet1 = Unet(dim=128, image_embed_dim=512, cond_dim=128, channels=3, dim_mults=(1, 2, 4, 8)).cuda()

unet2 = Unet(dim=16, image_embed_dim=512, cond_dim=128, channels=3, dim_mults=(1, 2, 4, 8, 16)).cuda()

decoder = Decoder(unet=(unet1, unet2), image_sizes=(128, 256), clip=clip, timesteps=100, image_cond_drop_prob=0.1, text_cond_drop_prob=0.5, condition_on_text_encodings=False).cuda()

for unet_number in (1, 2):
    loss = decoder(images, unet_number=unet_number)  # this can optionally be decoder(images, text) if you wish to
    # condition on the text encodings as well, though it was hinted in the paper it didn't do much
    loss.backward()

# do above for many steps

dalle2 = DALLE2(prior=diffusion_prior,decoder=decoder)
# classifier free guidance strength (> 1 would strengthen the condition)
images = dalle2(['a cute cat chasing after a squirrel'], cond_scale=2.)

# save our image (in this example, of size 256x256)
# plt.imshow(images)
# plt.savefig(images, str(uuid.uuid4()) + '_img.png')
```

**Comment by mu745511** [comment]
@LITDataScience  in order to save the predicted images you will have to do the following .
images = dalle2(['a cute cat chasing after a squirrel'], cond_scale=2.).detach().cpu().numpy() # detach it from cuda and convert it to numpy 
images = images.squeeze(axis = 0) # since the output is 4D and 0 axis is your batch size of 1 we can squeeze the axis .
images = np.transpose(images, (1,2,0)) # shifting the channel to last axis as pytorch takes image input as batch_size x Channel x Image width x Image Height . 
plt.imsave("dalle2_output.png",images) # finally save your prediction .

**Comment by lucidrains** [comment]
@LITDataScience you can just set `return_pil_images = True` on invoking `dalle2` https://github.com/lucidrains/DALLE2-pytorch/blob/main/dalle2_pytorch/dalle2_pytorch.py#L2243 Pillow image instances have a `save` method that you can then call to store the file anywhere

you'll still have to train the neural network first on a lot of data though before you will see any reasonable images

**Comment by kurtbomya** [comment]
Hey @lucidrains , do you've a guide on how you go about training the neural network first? I've been reading your READMEs and attempting to get started, as well as searching the Discord you've setup. I'm curious how many cycles we're talking for the training network (1,000 cycles? 60,000 cycles?) And where do you get the data from? is the Dalle2 README you have on the front of this repository enough or do I need to import more data? I was looking for a collection of data in your repository but didn't find any. So far my output has been a lot of this: 
![image](https://user-images.githubusercontent.com/23176626/198718278-34db2008-857b-44b5-bb61-4f0bf23c69eb.png)

**Comment by qianghuang84** [comment]
good
